src/ucp/core/ucp_request.c

```diff
@@ -461,51 +400,73 @@ UCS_PROFILE_FUNC(ucs_status_t, ucp_request_memory_reg,
+            goto err;
+        }
         ucp_trace_req(req, "mem reg md_map 0x%" PRIx64 "/0x%" PRIx64,
-                      state->dt.contig.md_map, md_map);
+                      state->dt.contig.memh->md_map, reg_md_map);
         break;
     case UCP_DATATYPE_IOV:
+        ucs_assert(!(flags & UCT_MD_MEM_FLAG_HIDE_ERRORS));
         iovcnt = state->dt.iov.iovcnt;
         iov    = buffer;
-        dt_reg = ((state->dt.iov.dt_reg == NULL) ?
-                  ucs_calloc(iovcnt, sizeof(*dt_reg), "iov_dt_reg") :
-                  state->dt.iov.dt_reg);
-        if (NULL == dt_reg) {
-            status = UCS_ERR_NO_MEMORY;
-            goto err;
+        if (ucs_unlikely(state->dt.iov.memhs != NULL)) {
+            memhs = state->dt.iov.memhs;
+            update = 1;
+        } else {
+            memhs = ucs_calloc(iovcnt, sizeof(*memhs), "iov_memhs");
+            if (NULL == memhs) {
+                status = UCS_ERR_NO_MEMORY;
+                goto err;
+            }
         }
+
         for (iov_it = 0; iov_it < iovcnt; ++iov_it) {
-            if (iov[iov_it].length) {
-                status = ucp_mem_rereg_mds(context, md_map, iov[iov_it].buffer,
-                                           iov[iov_it].length, flags, NULL,
-                                           mem_type, NULL, dt_reg[iov_it].memh,
-                                           &dt_reg[iov_it].md_map);
-                if (status != UCS_OK) {
-                    /* unregister previously registered memory */
-                    ucp_request_dt_dereg(context, dt_reg, iov_it, req);
-                    ucs_free(dt_reg);
-                    goto err;
+            if (update) {
+                if (ucs_test_all_flags(memhs[iov_it]->md_map, reg_md_map)) {
+                    continue;
                 }
-                ucp_trace_req(req,
-                              "mem reg iov %ld/%ld md_map 0x%" PRIx64
-                              "/0x%" PRIx64,
-                              iov_it, iovcnt, dt_reg[iov_it].md_map, md_map);
+                status = ucp_memh_update(context, reg_md_map, flags,
+                                         memhs[iov_it]);
+            } else {
+                status = ucp_memh_get(context, iov[iov_it].buffer,
+                                      iov[iov_it].length, mem_type, reg_md_map,
+                                      flags, &memhs[iov_it]);
             }
+            if (status != UCS_OK) {
+                /* unregister previously registered memory */
+                ucp_request_dt_dereg(context, memhs, iov_it, req);
+                ucs_free(memhs);
+                goto err;
+            }
+
+            ucp_trace_req(req, "mem reg iov %ld/%ld md_map 0x%" PRIx64
+                          "/0x%" PRIx64 " rereg %d", iov_it, iovcnt,
+                          memhs[iov_it]->md_map, reg_md_map, update);
         }
-        state->dt.iov.dt_reg = dt_reg;
+        state->dt.iov.memhs  = memhs;
```

Comment:
1. maybe just assign it in line 439?
2. maybe ucp_memh_update could include checking ig memh is NULL and if yes call ucp_memh_get. it means it would accept ucp_mem_h* instead of ucp_mem_h